pkg/controllers/clusters/clusterlifecycle/clusterlifecycle_controller.go

```diff
@@ -0,0 +1,352 @@
+/*
+Copyright 2021 The Clusternet Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterlifecycle
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	clusterapi "github.com/clusternet/clusternet/pkg/apis/clusters/v1beta1"
+	clusternetClientSet "github.com/clusternet/clusternet/pkg/generated/clientset/versioned"
+	clusterinformers "github.com/clusternet/clusternet/pkg/generated/informers/externalversions/clusters/v1beta1"
+	clusterlisters "github.com/clusternet/clusternet/pkg/generated/listers/clusters/v1beta1"
+)
+
+const (
+	DefaultClusterMonitorGracePeriod = 9 * time.Minute
+
+	// ClusterNotReadyThreshold indicate the threshold for cluster not ready period
+	ClusterNotReadyThreshold = 3
+)
+
+// Controller is a controller that manages cluster's lifecycle
+type Controller struct {
+	ctx context.Context
+
+	clusternetClient clusternetClientSet.Interface
+
+	// workqueue is a rate limited work queue. This is used to queue work to be
+	// processed instead of performing it as soon as a change happens. This
+	// means we can ensure we only process a fixed amount of resources at a
+	// time, and makes it easy to ensure we are never processing the same item
+	// simultaneously in two different workers.
+	workqueue workqueue.RateLimitingInterface
+
+	clusterLister clusterlisters.ManagedClusterLister
+	clusterSynced cache.InformerSynced
+
+	recorder record.EventRecorder
+}
+
+func NewController(ctx context.Context, clusternetClient clusternetClientSet.Interface,
+	clusterInformer clusterinformers.ManagedClusterInformer, recorder record.EventRecorder) *Controller {
+	c := &Controller{
+		ctx:              ctx,
+		clusternetClient: clusternetClient,
+		workqueue:        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "managedCluster"),
+		clusterLister:    clusterInformer.Lister(),
+		clusterSynced:    clusterInformer.Informer().HasSynced,
+		recorder:         recorder,
+	}
+
+	clusterInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addCluster,
+		UpdateFunc: c.updateCluster,
+		DeleteFunc: c.deleteCluster,
```

Comment:
Seems there is no need to handle cluster deleting. The `ManagedCluster` object gets deleted, no more updates (including conditions, status) are needed.

Reply:
Cluster deleting is a long process,  user may found deletion is hanging because cluster unknown or other problems. so status monitoring maybe useful here.
But i notice that subscription controller don't handle cluster deletion now, so we also can just ignore delete events here.

Reply:
`ManagedCluster` is just a description info on cluster, which does not block anything on deleting.

> But i notice that subscription controller don't handle cluster deletion now, so we can also ignore delete events here.

If it is not needed, then please do not add it.

Reply:
> user may found deletion is hanging because cluster unknown or other problems. so status monitoring maybe useful here.

`ManagedCluster` could be deleted immediately.

Reply:
addressed